src/main/scala/tech/cryptonomic/conseil/Lorre.scala

```diff
@@ -219,7 +202,32 @@ object Lorre extends App with TezosErrors with LazyLogging with LorreAppConfig w
     def processBakingAndEndorsingRights(fetchingResults: tezosNodeOperator.BlockFetchingResults): Future[Unit] = {
       import cats.implicits._
       val bh = fetchingResults.map(_._1.data.hash)
-      (tezosNodeOperator.getBatchBakingRights(bh), tezosNodeOperator.getBatchEndorsingRights(bh)).mapN {
+
+      def safeHead[A](list : List[A]): Option[A] = {
+        list match {
+          case Nil => None
+          case x :: _ => Some(x)
+        }
+      }
+
+      val cycle = safeHead(fetchingResults).flatMap{
+        _._1.data.metadata match {
+          case GenesisMetadata => None
+          case BlockHeaderMetadata(_,_,_,_,_,level) => Some(level.cycle)
+        }
+      }
+
+      val governancePeriod = safeHead(fetchingResults).flatMap{
+        _._1.data.metadata match {
+          case GenesisMetadata => None
+          case BlockHeaderMetadata(_,_,_,_,_,level) => Some(level.voting_period)
+        }
+      }
+
+      val blockHashesWithCycleAndGovernancePeriod = bh.map(hash => (cycle, governancePeriod, hash))
```

Comment:
I'm not sure I follow, but this way it looks like for the whole batch of block-hashes, we always assign the cycle and voting period of the first block in the batch.
Naively I'd think that each block would need to be assigned to his own level.cycle/level.voting_period.
Hence cycle, and governancePeriod might be turned into functions to combine when mapping over all results, to get the triple.

Reply:
You're absolutely right, I'll refactor this so that each block hash is coupled with it's own cycle/period.